pycargoebuild/cargo.py

```diff
@@ -63,11 +63,29 @@ class GitCrate(Crate):
 
     @property
     def download_url(self) -> str:
-        return f"{self.repository}/archive/{self.commit}.tar.gz"
+        if self.repository.startswith("https://github.com/"):
+            return f"{self.repository}/archive/{self.commit}.tar.gz"
+        if self.repository.startswith("https://gitlab."):
+            return (f"{self.repository}/-/archive/{self.commit}/"
+                    f"{self.repo_name}-{self.commit}.tar.gz")
+        raise RuntimeError(f"Unsupported git crate source: "
+                           f"{self.repository}")
 
     @property
     def filename(self) -> str:
-        return f"{self.repository.rpartition('/')[2]}-{self.commit}.gh.tar.gz"
+        return f"{self.repo_name}-{self.commit}{self.repo_ext}.tar.gz"
+
+    @property
+    def repo_ext(self) -> str:
+        if self.repository.startswith("https://github.com/"):
+            return ".gh"
+        if self.repository.startswith("https://gitlab.com/"):
+            return ".gl"
+        return ""
```

Comment:
Shouldn't this `raise` too? I mean, technically `download_url` will raise anyway, but having symmetry would help when someone adds a new type in the future, and forgets to update this.

Reply:
I meant to write it as a wildcard match to something other than `git{hub,lab}.com`.
Your point made sense to me, so I fixed it to match the behavior with download_url().